File: scripts/optimize_images.py
#!/usr/bin/env python
"""
Optimiza imágenes JPEG/PNG del frontend a WebP usando Pillow.

Uso (desde la raíz del repo):
    python scripts/optimize_images.py

Requiere Pillow (ya incluido en backend/requirements.txt).
Activá el venv del backend antes de correrlo o pasale el python adecuado.

Estrategia:
    - Convierte cada archivo .jpg/.jpeg/.png a .webp con calidad 75 manteniendo
      el original como fallback (para navegadores sin WebP, hoy >97% support).
    - Imprime el ahorro de tamaño por archivo.

En el HTML/CSS, usar:
    <picture>
      <source srcset="/img/background/foo.webp" type="image/webp" />
      <img src="/img/background/foo.jpg" alt="..." />
    </picture>
"""
from __future__ import annotations

import sys
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    sys.exit(
        "Pillow no está instalado. Activá el venv del backend o instala con `pip install Pillow`."
    )

ROOT = Path(__file__).resolve().parent.parent
IMG_DIRS = [
    ROOT / "client/public/img/background",
]
QUALITY = 75
VALID_EXT = {".jpg", ".jpeg", ".png"}


def main() -> None:
    total_in = total_out = count = 0
    for d in IMG_DIRS:
        if not d.exists():
            print(f"  [!] {d} no existe, salto")
            continue
        for path in d.iterdir():
            if path.suffix.lower() not in VALID_EXT:
                continue
            out_path = path.with_suffix(".webp")
            try:
                with Image.open(path) as im:
                    # Convertir a RGB si tiene canal alpha PNG sin transparencia visible
                    if im.mode in ("RGBA", "P"):
                        im = im.convert("RGB")
                    im.save(out_path, "WEBP", quality=QUALITY)
                in_size = path.stat().st_size
                out_size = out_path.stat().st_size
                pct = (in_size - out_size) / in_size * 100
                print(
                    f"  {path.name:<40} {in_size // 1024:>5} KB -> {out_size // 1024:>5} KB  (-{pct:.1f}%)"
                )
                total_in += in_size
                total_out += out_size
                count += 1
            except Exception as exc:  # noqa: BLE001
                print(f"  [!] Falló {path.name}: {exc}")

    if count == 0:
        print("  (no se encontraron imágenes para optimizar)")
        return

    saved = (total_in - total_out) / 1024
    pct = (total_in - total_out) / total_in * 100
    print(f"\nResumen: {count} imágenes optimizadas, {saved:.0f} KB ahorrados (-{pct:.1f}%).")


if __name__ == "__main__":
    main()

File: client/scripts/optimize-images.mjs
#!/usr/bin/env node
/**
 * Optimiza imágenes JPEG/PNG del `public/img/` a WebP.
 *
 * Uso (desde la raíz de `client/`):
 *     node scripts/optimize-images.mjs
 *
 * Requiere `sharp` (instalar con `pnpm add -D sharp` o `npm i -D sharp`).
 * Si no está disponible, imprime instrucciones y termina sin fallar el build.
 *
 * Estrategia:
 *   - Recorre `public/img/background/` y crea variantes `.webp` (calidad 75)
 *     manteniendo originales como fallback.
 *   - Imprime el ahorro de tamaño por archivo.
 *
 * En el HTML/CSS, usar:
 *   <picture>
 *     <source srcset="/img/background/foo.webp" type="image/webp" />
 *     <img src="/img/background/foo.jpg" alt="..." />
 *   </picture>
 */
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');
const IMG_DIRS = [path.join(ROOT, 'public/img/background')];
const QUALITY = 75;
const VALID_EXT = new Set(['.jpg', '.jpeg', '.png']);

async function main() {
  let sharp;
  try {
    ({ default: sharp } = await import('sharp'));
  } catch {
    console.error(
      '\n[optimize-images] El paquete "sharp" no está instalado.\n' +
      'Para optimizar las imágenes a WebP ejecutá:\n\n' +
      '    pnpm add -D sharp\n' +
      '    node scripts/optimize-images.mjs\n\n' +
      'Alternativamente, podés convertir manualmente usando cwebp:\n' +
      '    cwebp -q 75 entrada.jpg -o salida.webp\n'
    );
    process.exit(0);
  }

  let total = { input: 0, output: 0, count: 0 };

  for (const dir of IMG_DIRS) {
    try {
      const files = await fs.readdir(dir);
      for (const file of files) {
        const ext = path.extname(file).toLowerCase();
        if (!VALID_EXT.has(ext)) continue;

        const inputPath = path.join(dir, file);
        const outputPath = path.join(dir, file.replace(ext, '.webp'));

        const inputSize = (await fs.stat(inputPath)).size;

        await sharp(inputPath)
          .webp({ quality: QUALITY })
          .toFile(outputPath);

        const outputSize = (await fs.stat(outputPath)).size;
        const saved = inputSize - outputSize;
        const pct = ((saved / inputSize) * 100).toFixed(1);
        console.log(
          `  ${file.padEnd(40)} ${(inputSize / 1024).toFixed(0).padStart(5)} KB → ${(outputSize / 1024).toFixed(0).padStart(5)} KB  (-${pct}%)`,
        );
        total.input += inputSize;
        total.output += outputSize;
        total.count += 1;
      }
    } catch (err) {
      console.warn(`  ⚠ No se pudo leer ${dir}: ${err.message}`);
    }
  }

  if (total.count === 0) {
    console.log('  (no se encontraron imágenes para optimizar)');
    return;
  }

  const savedKB = ((total.input - total.output) / 1024).toFixed(0);
  const pct = (((total.input - total.output) / total.input) * 100).toFixed(1);
  console.log(
    `\nResumen: ${total.count} imágenes optimizadas, ${savedKB} KB ahorrados (-${pct}%).`,
  );
}

main().catch((err) => {
  console.error('Error optimizando imágenes:', err);
  process.exit(1);
});
